internal/dcache/cluster_manager/cluster_manager_impl.go

```diff
@@ -508,7 +510,130 @@ func (cmi *ClusterManagerImpl) updateStorageClusterMapIfRequired() {
 
 }
 
-func (cmi *ClusterManagerImpl) updateMVList(clusterMapRVMap map[string]dcache.RawVolume, clusterMapMVMap map[string]dcache.MirroredVolume) {
+func updateMVList(rvMap map[string]dcache.RawVolume, existingMVMap map[string]dcache.MirroredVolume, NumReplicas int, MvsPerRv int) map[string]dcache.MirroredVolume {
+
+	// Local types
+	type rv struct {
+		rvName string
+		slots  int //MvsPerRv
+	}
+
+	type node struct {
+		nodeId string
+		rvs    []rv
+	}
+
+	nodeToRvs := make(map[string]node)
+	// Populate the RV struct and node struct
+	for rvId, rvInfo := range rvMap {
+		if _, exists := nodeToRvs[rvInfo.NodeId]; exists {
+			// If the node already exists, append the RV to its list
+			node := nodeToRvs[rvInfo.NodeId]
+			node.rvs = append(node.rvs, rv{
+				rvName: rvId,
+				slots:  MvsPerRv,
+			})
+			nodeToRvs[rvInfo.NodeId] = node
+		} else {
+			// Create a new node and add the RV to it
+			nodeToRvs[rvInfo.NodeId] = node{
+				nodeId: rvInfo.NodeId,
+				rvs:    []rv{{rvName: rvId, slots: MvsPerRv}},
+			}
+		}
+	}
+
+	// Phase 1 : Check and update slots of Rv's used in existing MVs
+	for _, rVWithStateMap := range existingMVMap {
+		// Should we check for the state of the RV?
+		for rvName := range rVWithStateMap.RVWithStateMap {
+			nodeId := rvMap[rvName].NodeId
+			for i := range len(nodeToRvs[nodeId].rvs) {
+				if nodeToRvs[nodeId].rvs[i].rvName == rvName {
+					nodeToRvs[nodeId].rvs[i].slots--
+				}
+			}
+		}
+	}
+
+	// Phase 2 : Generate new MVs
+	for {
+		// Stored nodes in a slice as its wasy to shuffle
+		var availableNodes []node
+		for _, n := range nodeToRvs {
+			availableNodes = append(availableNodes, n)
+
+		}
+
+		max := (len(rvMap) * MvsPerRv) / NumReplicas
+		// End of MV generation if we have enough MVs or if we have exhausted all the nodes
+		if len(availableNodes) < NumReplicas || len(existingMVMap) >= max {
+			break
+		}
```

Comment:
by using the correct slot values we shouldn't need this.
Infact we can instead assert for these.